File: src/backend/nano_tob.py
"""
nano_tob.py — Thread Object Broker 后端客户端
供后端节点使用的 TOB API。

使用前需先调用 initializeTOBM() 初始化 Thread Object Broker 机制。
线程 ID 通过环境变量 THREAD_ID 获取。
"""

import asyncio
import json
import os
import secrets
import string

import websockets


# ── nanoid 实现 ──────────────────────────────────────────────────────

def _nanoid(size=21):
    alphabet = string.ascii_letters + string.digits + "_-"
    return ''.join(secrets.choice(alphabet) for _ in range(size))


def _make_ins_id():
    return f"Ins-{_nanoid()}"


# ── TOB 对象 ─────────────────────────────────────────────────────────

class TOB:
    """TOB 本地代理对象"""
    def __init__(self, tob_id):
        self.id = tob_id
        self._methods = {}


# ── 客户端 ───────────────────────────────────────────────────────────

class _TOBNode:
    def __init__(self):
        self.ws = None
        self.thread_id = os.environ.get("THREAD_ID", "")
        self._pending = {}   # ins_id -> asyncio.Future
        self._tobs = {}      # tob_id -> TOB

    async def connect(self, url):
        self.ws = await websockets.connect(url)
        asyncio.create_task(self._listen())

    async def close(self):
        """关闭连接并清理本地状态"""
        self._tobs.clear()
        for future in self._pending.values():
            if not future.done():
                future.set_exception(RuntimeError("连接已关闭"))
        self._pending.clear()
        if self.ws is not None:
            try:
                await self.ws.close()
            except Exception:
                pass
            self.ws = None

    async def _listen(self):
        """持续监听服务器消息"""
        try:
            async for message in self.ws:
                try:
                    data = json.loads(message)
                except json.JSONDecodeError:
                    continue
                ins_type = data.get("type", "")
                ins_id = data.get("id", "")
                params = data.get("params", {})

                if ins_type == "callTOBMethod":
                    # 服务器要求我们执行本地方的方法
                    asyncio.create_task(self._handle_incoming_call(ins_id, params))
                else:
                    # 响应我们发出的请求
                    future = self._pending.pop(ins_id, None)
                    if future and not future.done():
                        future.set_result(params)
        except websockets.ConnectionClosed:
            pass

    async def _handle_incoming_call(self, ins_id, params):
        tob_id = params.get("tob_id", "")
        method = params.get("method", "")
        method_params = params.get("params", [])

        tob = self._tobs.get(tob_id)
        if not tob:
            await self._send_return(ins_id, error=f"TOB {tob_id} 在本地未找到")
            return

        delegate = tob._methods.get(method)
        if not delegate:
            await self._send_return(ins_id, error=f"TOB {tob_id} 上未挂载方法 '{method}'")
            return

        try:
            if asyncio.iscoroutinefunction(delegate):
                result = await delegate(*method_params)
            else:
                result = delegate(*method_params)
            await self._send_return(ins_id, result=result)
        except Exception as e:
            await self._send_return(ins_id, error=str(e))

    async def _send_return(self, ins_id, result=None, error=None):
        await self.ws.send(json.dumps({
            "thread_id": self.thread_id,
            "id": ins_id,
            "type": "callTOBMethodReturn",
            "params": {"result": result, "error": error}
        }, ensure_ascii=False))

    async def _request(self, msg_type, params):
        ins_id = _make_ins_id()
        loop = asyncio.get_running_loop()
        future = loop.create_future()
        self._pending[ins_id] = future

        await self.ws.send(json.dumps({
            "thread_id": self.thread_id,
            "id": ins_id,
            "type": msg_type,
            "params": params
        }, ensure_ascii=False))

        return await future


# ── 模块级单例 ────────────────────────────────────────────────────────

_client = None


async def initializeTOBM(url="ws://127.0.0.1:8000/ws"):
    """初始化 Thread Object Broker 机制。必须先调用此函数。"""
    global _client
    _client = _TOBNode()
    await _client.connect(url)


async def closeTOBM():
    """关闭 Thread Object Broker 机制。关闭后可通过 initializeTOBM 重新初始化。"""
    global _client
    if _client is not None:
        try:
            await _client.close()
        finally:
            _client = None


def _ensure_client():
    if _client is None:
        raise RuntimeError("Thread Object Broker 未初始化。请先调用 initializeTOBM(url)。")


# ── 公开 API ─────────────────────────────────────────────────────────

async def createTOB():
    """创建一个 TOB，返回 TOB 对象"""
    _ensure_client()
    result = await _client._request("createTOB", {})
    tob_id = result["tob_id"]
    tob = TOB(tob_id)
    _client._tobs[tob_id] = tob
    return tob


async def nameTOB(tob, name):
    """给 TOB 命名"""
    _ensure_client()
    result = await _client._request("nameTOB", {"tob_id": tob.id, "name": name})
    return result.get("success", False)


async def getTOB(tob_id):
    """通过 ID 获取 TOB，不存在返回 None"""
    _ensure_client()
    result = await _client._request("getTOB", {"tob_id": tob_id})
    if result.get("tob_id"):
        tid = result["tob_id"]
        tob = _client._tobs.get(tid)
        if tob is None:
            tob = TOB(tid)
            _client._tobs[tid] = tob
        return tob
    return None


async def waitTOB(tob_id, timeout=0):
    """
    等待指定 ID 的 TOB 被创建。
    timeout 为毫秒，0 表示无限等待。超时返回 None。
    """
    _ensure_client()
    result = await _client._request("waitTOB", {"tob_id": tob_id, "timeout": timeout})
    if result.get("tob_id"):
        tid = result["tob_id"]
        tob = _client._tobs.get(tid)
        if tob is None:
            tob = TOB(tid)
            _client._tobs[tid] = tob
        return tob
    return None


async def waitNamedTOB(name, timeout=0):
    """
    等待指定名称的 TOB 被创建。
    timeout 为毫秒，0 表示无限等待。超时返回 None。
    """
    _ensure_client()
    result = await _client._request("waitNamedTOB", {"name": name, "timeout": timeout})
    if result.get("tob_id"):
        tid = result["tob_id"]
        tob = _client._tobs.get(tid)
        if tob is None:
            tob = TOB(tid)
            _client._tobs[tid] = tob
        return tob
    return None


async def callTOBMethod(tob, method, params):
    """调用 TOB 上的远程方法"""
    _ensure_client()
    result = await _client._request("callTOBMethod", {
        "tob_id": tob.id,
        "method": method,
        "params": params
    })
    if result.get("error"):
        raise Exception(result["error"])
    return result.get("result")


def mountTOBMethod(tob, method, delegate):
    """
    在 TOB 上挂载本地方法。纯本地操作，不通知服务器。
    当服务器转发 callTOBMethod 到此节点时，会调用该方法。
    """
    tob._methods[method] = delegate


async def forgetTOB(tob_id):
    """删除指定 ID 的 TOB"""
    _ensure_client()
    result = await _client._request("forgetTOB", {"tob_id": tob_id})
    if result.get("success"):
        _client._tobs.pop(tob_id, None)
        return True
    return False

File: src/frontend/public/nano_tob.js
/**
 * nano_tob.js — Thread Object Broker 前端客户端
 * 供前端节点使用的 TOB API。
 *
 * 使用前需先调用 initializeTOBM() 初始化 Thread Object Broker 机制。
 * 线程 ID 通过 URL query string 的 thread_id 参数获取。
 */

// ── nanoid 实现 ──────────────────────────────────────────────────────

function nanoid(size = 21) {
    const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_-';
    const bytes = new Uint8Array(size);
    crypto.getRandomValues(bytes);
    let id = '';
    for (let i = 0; i < size; i++) {
        id += alphabet[bytes[i] % alphabet.length];
    }
    return id;
}

function makeInsId() {
    return 'Ins-' + nanoid();
}

// ── TOB 对象 ─────────────────────────────────────────────────────────

class TOB {
    constructor(tobId) {
        /** @type {string} */
        this.id = tobId;
        /** @type {Object<string, Function>} */
        this._methods = {};
    }
}

// ── 客户端 ───────────────────────────────────────────────────────────

class TOBNode {
    constructor() {
        /** @type {WebSocket|null} */
        this.ws = null;
        this.threadId = getThreadId();
        /** @type {Object<string, {resolve: Function, reject: Function}>} */
        this._pending = {};
        /** @type {Object<string, TOB>} */
        this._tobs = {};
    }

    /**
     * 连接到系统通信服务
     * @param {string} url
     * @returns {Promise<void>}
     */
    connect(url) {
        return new Promise((resolve, reject) => {
            this.ws = new WebSocket(url);
            this.ws.onopen = () => resolve();
            this.ws.onerror = (err) => reject(err);
            this.ws.onclose = () => {
                for (const id of Object.keys(this._pending)) {
                    this._pending[id].reject(new Error('连接已关闭'));
                }
                this._pending = {};
            };
            this.ws.onmessage = (event) => this._onMessage(JSON.parse(event.data));
        });
    }

    /** 关闭连接并清理本地状态 */
    close() {
        this._tobs = {};
        for (const id of Object.keys(this._pending)) {
            this._pending[id].reject(new Error('连接已关闭'));
        }
        this._pending = {};
        if (this.ws) {
            this.ws.close();
            this.ws = null;
        }
    }

    /**
     * 处理服务器消息
     * @param {{type: string, id: string, params: Object}} data
     */
    _onMessage(data) {
        const { type, id, params } = data;

        if (type === 'callTOBMethod') {
            this._handleIncomingCall(id, params);
        } else {
            const pending = this._pending[id];
            if (pending) {
                delete this._pending[id];
                pending.resolve(params);
            }
        }
    }

    /**
     * 处理服务器转发来的方法调用
     * @param {string} insId
     * @param {{tob_id: string, method: string, params: Array}} params
     */
    async _handleIncomingCall(insId, params) {
        const { tob_id, method, params: methodParams } = params;
        const tob = this._tobs[tob_id];

        if (!tob || !tob._methods[method]) {
            this._sendReturn(insId, null, `TOB ${tob_id} 上未挂载方法 '${method}'`);
            return;
        }

        try {
            const result = await tob._methods[method](...(methodParams || []));
            this._sendReturn(insId, result, null);
        } catch (e) {
            this._sendReturn(insId, null, e.message || String(e));
        }
    }

    /**
     * 发送方法调用结果回服务器
     * @param {string} insId
     * @param {*} result
     * @param {string|null} error
     */
    _sendReturn(insId, result, error) {
        if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;
        this.ws.send(JSON.stringify({
            thread_id: this.threadId,
            id: insId,
            type: 'callTOBMethodReturn',
            params: { result: result, error: error }
        }));
    }

    /**
     * 发送请求并返回 Promise
     * @param {string} type
     * @param {Object} params
     * @returns {Promise<Object>}
     */
    _request(type, params) {
        return new Promise((resolve, reject) => {
            if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
                reject(new Error('系统通信未连接'));
                return;
            }
            const insId = makeInsId();
            this._pending[insId] = { resolve, reject };
            this.ws.send(JSON.stringify({
                thread_id: this.threadId,
                id: insId,
                type: type,
                params: params
            }));
        });
    }
}

// ── 模块级单例 ────────────────────────────────────────────────────────

/** @type {TOBNode|null} */
let _client = null;

/**
 * 初始化 Thread Object Broker 机制。必须先调用此函数。
 * @param {string} url
 * @returns {Promise<void>}
 */
async function initializeTOBM(url = 'ws://127.0.0.1:8000/ws') {
    _client = new TOBNode();
    await _client.connect(url);
}

async function closeTOBM() {
    if (_client) {
        try {
            await _client.close();
        } finally {
            _client = null;
        }
    }
}

/**
 * 获取当前线程 ID
 * @returns {string}
 */
function getThreadId() {
    return new URLSearchParams(window.location.search).get('thread_id') || '';
}

function _ensureClient() {
    if (!_client) {
        throw new Error('Thread Object Broker 未初始化。请先调用 initializeTOBM(url)。');
    }
}

// ── 公开 API ─────────────────────────────────────────────────────────

/**
 * 创建一个 TOB，返回 TOB 对象
 * @returns {Promise<TOB>}
 */
async function createTOB() {
    _ensureClient();
    const result = await _client._request('createTOB', {});
    const tob = new TOB(result.tob_id);
    _client._tobs[result.tob_id] = tob;
    return tob;
}

/**
 * 给 TOB 命名
 * @param {TOB} tob
 * @param {string} name
 * @returns {Promise<boolean>}
 */
async function nameTOB(tob, name) {
    _ensureClient();
    const result = await _client._request('nameTOB', { tob_id: tob.id, name: name });
    return result.success || false;
}

/**
 * 通过 ID 获取 TOB，不存在返回 null
 * @param {string} tobId
 * @returns {Promise<TOB|null>}
 */
async function getTOB(tobId) {
    _ensureClient();
    const result = await _client._request('getTOB', { tob_id: tobId });
    if (result.tob_id) {
        let tob = _client._tobs[result.tob_id];
        if (!tob) {
            tob = new TOB(result.tob_id);
            _client._tobs[result.tob_id] = tob;
        }
        return tob;
    }
    return null;
}

/**
 * 等待指定 ID 的 TOB 被创建。
 * @param {string} tobId
 * @param {number} timeout 毫秒，0 表示无限等待
 * @returns {Promise<TOB|null>}
 */
async function waitTOB(tobId, timeout = 0) {
    _ensureClient();
    const result = await _client._request('waitTOB', { tob_id: tobId, timeout: timeout });
    if (result.tob_id) {
        let tob = _client._tobs[result.tob_id];
        if (!tob) {
            tob = new TOB(result.tob_id);
            _client._tobs[result.tob_id] = tob;
        }
        return tob;
    }
    return null;
}

/**
 * 等待指定名称的 TOB 被创建。
 * @param {string} name
 * @param {number} timeout 毫秒，0 表示无限等待
 * @returns {Promise<TOB|null>}
 */
async function waitNamedTOB(name, timeout = 0) {
    _ensureClient();
    const result = await _client._request('waitNamedTOB', { name: name, timeout: timeout });
    if (result.tob_id) {
        let tob = _client._tobs[result.tob_id];
        if (!tob) {
            tob = new TOB(result.tob_id);
            _client._tobs[result.tob_id] = tob;
        }
        return tob;
    }
    return null;
}

/**
 * 调用 TOB 上的远程方法
 * @param {TOB} tob
 * @param {string} method
 * @param {Array} params
 * @returns {Promise<*>}
 */
async function callTOBMethod(tob, method, params) {
    _ensureClient();
    const result = await _client._request('callTOBMethod', {
        tob_id: tob.id,
        method: method,
        params: params
    });
    if (result.error) {
        throw new Error(result.error);
    }
    return result.result;
}

/**
 * 在 TOB 上挂载本地方法。纯本地操作，不通知服务器。
 * @param {TOB} tob
 * @param {string} method
 * @param {Function} delegate
 */
function mountTOBMethod(tob, method, delegate) {
    tob._methods[method] = delegate;
}

/**
 * 删除指定 ID 的 TOB
 * @param {string} tobId
 * @returns {Promise<boolean>}
 */
async function forgetTOB(tobId) {
    _ensureClient();
    const result = await _client._request('forgetTOB', { tob_id: tobId });
    if (result.success) {
        delete _client._tobs[tobId];
        return true;
    }
    return false;
}
